### ISSUE TITLE
[BUG] Rank-3 tensor slices into rank-3 DFB blocks generate broken datamovement: wrong results (1-tile blocks), device deadlock (multi-tile blocks)

### ISSUE THREAD

**Issue by ayerofieiev-tt** [tenstorrent/tt-lang#898]
## Description

Indexing a rank-3 tensor with rank-3 slices into rank-3 dataflow-buffer blocks (e.g. `x[b:b+1, rt:rt+1, c0:c1]` into `make_dataflow_buffer_like(x, shape=(1, 1, WC))`) generates broken **datamovement** kernels in the light wheel `1.1.7.dev20260803+light`:

1. **Single-tile blocks `(1, 1, 1)`**: the kernel runs but produces wrong results — data is read/placed incorrectly (in our case an intra-tile ROW broadcast of the block behaved as identity, i.e. rows 1–31 of the tile never saw the broadcast operand's data).
2. **Multi-tile blocks `(1, 1, 12)`**: the program **deadlocks on device** (host blocks forever in `ttnn.to_torch`). This happens even with *no broadcast in the compute at all*, and even on a single core with `grid=(1, 1)`, so it is not a work-distribution or bcast/DST issue.

Evidence that the bug is in the generated DM kernels, not compute:

- The **compute** kernel content hash was *identical* between the broken rank-3 formulation and the working rank-2 formulation of the same op (`ttlang_kernel_compute_831446fe` in both) — only the read/write kernel hashes differed.
- Standalone 2D probes of `ttl.block.broadcast` (row `dims=[0]`, col `dims=[1]`, add and mul paths, equal-shape intra-tile) all pass on the same setup, so the LLK/bcast path is fine.
- Rewriting the same op with **scalar leading-dim indexing into rank-2 blocks** (`x[b, rt:rt+1, c0:c1]` into `shape=(1, WC)` DFBs — the idiom documented in `_make_tensor_slice`) is fully correct and runs at expected performance for both 1-tile and 12-tile-wide blocks, on 1 core and on the full grid.

Note the 3D tutorial example (`examples/eltwise_add_3d.py`) may not catch this: it uses identical DFB/tensor treatment for all operands; the failure shows up when blocks/slices differ per operand (here: activation rows `[b:b+1, rt:rt+1, c0:c1]` alongside a per-batch parameter row `[b:b+1, 0:1, c0:c1]`) and clearly with block width > 1 tile.

## To Reproduce

1. Install the light wheel `tt_lang-1.1.7.dev20260803+light` (cp310) against an external tt-metal build (main @ `240a20cc367`, Blackhole p150).
2. Run the code sample below with `MODE=mismatch` → wrong results.
3. Run with `MODE=hang` → host never returns from `ttnn.to_torch` (device deadlock; needs `tt-smi -r` afterwards).
4. Change `RANK3 = False` (rank-2 blocks, scalar batch index — only the copy/DFB lines differ) → both modes pass.

## Expected Behavior

`out[b, r, c] = x[b, r, c] * (1 + scale[b, c]) + shift[b, c]`, with `scale`/`shift` rows broadcast over the 32 rows of each tile; and no deadlock for 12-tile-wide blocks.

## Actual Behavior

- `MODE=mismatch` (rank-3, 1-tile blocks): `max_err ≈ 6.8` on randn data; error pattern is consistent with rows 1–31 receiving no broadcast data (`out == x` there).
- `MODE=hang` (rank-3, 12-tile blocks): device deadlock, host spins in `to_torch`. Reproduces with the broadcast removed from compute (`ob.store(ttl.add(xb, xb))`) and with `grid=(1, 1)`.
- `RANK3=False`: correct in both shapes (`max_err ≈ 0.11`, pure bf16 rounding at magnitude ~15), no hang, and ~1.5–2.5x faster than the equivalent composed ttnn ops.

## Code Sample

```python
# repro.py -- MODE=mismatch|hang, RANK3=True|False (False = working formulation)
import os
import torch
import ttl
import ttnn

MODE = os.environ.get("MODE", "mismatch")
RANK3 = os.environ.get("RANK3", "1") == "1"

TILE = 32
B = 1
T, D = (32, 32) if MODE == "mismatch" else (256, 384)
Tt, Dt = T // TILE, D // TILE
WC = Dt                 # 1 tile wide for mismatch, 12 tiles wide for hang
so_t, sh_t = 0, Dt      # scale = mod[..., 0:D], shift = mod[..., D:2D]
n_blocks = B * Tt


@ttl.operation(grid=(1, 1))
def modulate(x: ttnn.Tensor, mod: ttnn.Tensor, out: ttnn.Tensor) -> None:
    blk_shape = (1, 1, WC) if RANK3 else (1, WC)
    x_dfb = ttl.make_dataflow_buffer_like(x, shape=blk_shape, block_count=2)
    s_dfb = ttl.make_dataflow_buffer_like(mod, shape=blk_shape, block_count=2)
    h_dfb = ttl.make_dataflow_buffer_like(mod, shape=blk_shape, block_count=2)
    o_dfb = ttl.make_dataflow_buffer_like(out, shape=blk_shape, block_count=2)

    @ttl.compute()
    def compute():
        for blk in range(n_blocks):
            with (
                x_dfb.wait() as xb,
                s_dfb.wait() as sb,
                h_dfb.wait() as hb,
                o_dfb.reserve() as ob,
            ):
                row_dim = 1 if RANK3 else 0
                sbc = ttl.block.broadcast(sb, dims=[row_dim], shape=xb.shape)
                hbc = ttl.block.broadcast(hb, dims=[row_dim], shape=xb.shape)
                ob.store(ttl.add(ttl.add(ttl.mul(xb, sbc), xb), hbc))

    @ttl.datamovement()
    def read():
        for blk in range(n_blocks):
            b, rt = blk // Tt, blk % Tt
            with x_dfb.reserve() as xb, s_dfb.reserve() as sb, h_dfb.reserve() as hb:
                if RANK3:
                    tx = ttl.copy(x[b:b + 1, rt:rt + 1, 0:WC], xb)
                    ts = ttl.copy(mod[b:b + 1, 0:1, so_t:so_t + WC], sb)
                    th = ttl.copy(mod[b:b + 1, 0:1, sh_t:sh_t + WC], hb)
                else:
                    tx = ttl.copy(x[b, rt:rt + 1, 0:WC], xb)
                    ts = ttl.copy(mod[b, 0:1, so_t:so_t + WC], sb)
                    th = ttl.copy(mod[b, 0:1, sh_t:sh_t + WC], hb)
                tx.wait(); ts.wait(); th.wait()

    @ttl.datamovement()
    def write():
        for blk in range(n_blocks):
            b, rt = blk // Tt, blk % Tt
            with o_dfb.wait() as ob:
                if RANK3:
                    ttl.copy(ob, out[b:b + 1, rt:rt + 1, 0:WC]).wait()
                else:
                    ttl.copy(ob, out[b, rt:rt + 1, 0:WC]).wait()


device = ttnn.open_device(device_id=0)
try:
    torch.manual_seed(9)
    x_t = torch.randn((B, 1, T, D), dtype=torch.bfloat16)
    mod_t = torch.randn((B, 1, 1, 6 * D), dtype=torch.bfloat16)
    x3 = ttnn.reshape(ttnn.from_torch(x_t, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device), [B, T, D])
    mod3 = ttnn.reshape(ttnn.from_torch(mod_t, dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device), [B, 1, 6 * D])
    out3 = ttnn.from_torch(torch.zeros((B, T, D), dtype=torch.bfloat16),
                           dtype=ttnn.bfloat16, layout=ttnn.TILE_LAYOUT, device=device)
    modulate(x3, mod3, out3)
    r = ttnn.to_torch(out3).float()          # <-- hangs forever in MODE=hang with RANK3=1
    e = (x_t.float() * (1.0 + mod_t[..., 0:D].float()) + mod_t[..., D:2 * D].float()).squeeze(1)
    err = (r - e).abs().max().item()
    print(f"MODE={MODE} RANK3={RANK3} max_err={err:.4f}",
          "PASSED" if err < 0.15 else "MISMATCH")
finally:
    ttnn.close_device(device)
```

Run matrix observed:

| MODE | RANK3 | result |
|---|---|---|
| mismatch | 1 | `max_err=6.8191` MISMATCH |
| hang | 1 | device deadlock (also with bcast removed, and on `grid="full"`) |
| mismatch | 0 | PASSED |
| hang | 0 | PASSED (`max_err≈0.11`, bf16 rounding) |

## Environment

- **OS:** Ubuntu (Blackhole loudbox, p150 boards, fw bundle 19.8.1)
- **Python version:** 3.10.19
- **tt-lang version/commit:** wheel `tt_lang-1.1.7.dev20260803+light-cp310` from the S3 index (`pypi.eng.aws.tenstorrent.com/tt-lang/2026-08/`)
- **tt-metal:** external source build, main @ `240a20cc367` (build_Release; note this is ~1 month newer than the wheel's pinned TT_METAL_TAG — but the working rank-2 formulation rules out general API skew; generated kernels only emit deprecation warnings for `binary_op_init_common` / `add_tiles_init` / `unary_bcast_init`)
- **Hardware:** 1 Blackhole device (of 8 on the host), `ttnn.open_device(device_id=0)`

## Error Messages

No host-side errors in either failing mode: the mismatch mode completes silently with wrong data; the hang mode blocks in `ttnn.to_torch` with the device stuck (recoverable only via `tt-smi -r`). py-spy of the hung host process:

```
Thread (idle): "MainThread"
    __call__ (ttnn/decorators.py:650)
    to_torch (ttnn/operations/core.py:421)
    ...
```

**Comment by brnorris03** [comment]
No longer reproduces on main -- at `0e738abec` the rank-3 and rank-2 formulations of the reproducer generate identical reader, compute, and writer kernels, and on Blackhole `MODE=mismatch` passes (max_err 0.039) and `MODE=hang` completes (max_err 0.101). The fixing PR was not identified. Please reopen if it still reproduces on a current build. Closing.
